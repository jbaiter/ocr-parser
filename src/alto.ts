import { Detail, SaxEventType, Tag } from 'sax-wasm';
import {
  Dimensions,
  OcrBlock,
  OcrElement,
  OcrLine,
  OcrPage,
  OcrWord,
  Polygon,
  makeBlock,
  makeLine,
  makePage,
} from './model.js';
import {
  getParser,
  toReadableStream,
  resolveXmlEntities,
  getAttributeValue,
  getAllAttributes,
  logWithPosition,
} from './util.js';
import { ReferenceSizeCallback } from '.';

type ParserContext = {
  scaleFactor: number;
  elementStack: OcrElement[];
  sourceImageInformation?: OcrPage['imageSource'];
};

const NUMERIC_ATTRIBS = ['HEIGHT', 'WIDTH', 'HPOS', 'VPOS', 'WC'];

function getAttributes(
  data: Tag,
  names?: string[],
): Record<string, string | number> {
  return Object.fromEntries(
    data.attributes
      .filter((attr) => !names || names.includes(attr.name.value))
      .map((attr) => [
        attr.name.value,
        NUMERIC_ATTRIBS.includes(attr.name.value)
          ? Number.parseFloat(attr.value.value)
          : attr.value.value,
      ]),
  );
}

function getCoordinates(
  data: Tag,
): Partial<{ height: number; width: number; x: number; y: number }> {
  const attrs = getAttributes(data, ['HEIGHT', 'WIDTH', 'HPOS', 'VPOS']);
  return {
    height: attrs.HEIGHT as number | undefined,
    width: attrs.WIDTH as number | undefined,
    x: attrs.HPOS as number | undefined,
    y: attrs.VPOS as number | undefined,
  };
}

async function handlePage(
  eventIter: AsyncIterableIterator<[SaxEventType, Detail]>,
  ctx: ParserContext,
  tag: Tag,
  referenceSize?: Dimensions,
): Promise<OcrPage | null> {
  const dims = getCoordinates(tag);
  if (!dims.width || !dims.height) {
    throw new Error('Could not get Page dimensions');
  }
  const scaleFactorX = referenceSize ? referenceSize.width / dims.height! : 1;
  const scaleFactorY = referenceSize ? referenceSize.height / dims.width! : 1;
  if (scaleFactorX !== scaleFactorY) {
    console.warn(
      `Scale factors differ: x=${scaleFactorX} vs y=${scaleFactorY}, using X scale factor.`,
    );
  }
  ctx.scaleFactor = scaleFactorX;
  const page = makePage({
    features: [],
    width: dims.width * scaleFactorX,
    height: dims.height * scaleFactorX,
  });
  ctx.elementStack.push(page);

  const pageAttribs = getAllAttributes(tag);
  if ('PHYSICAL_IMG_NR' in pageAttribs) {
    page.physicalPageNumber = Number.parseInt(pageAttribs.PHYSICAL_IMG_NR);
  }
  if ('PRINTED_IMG_NR' in pageAttribs) {
    page.logicalPageNumber = pageAttribs.PRINTED_IMG_NR;
  }

  let openTags = 0;
  let result: IteratorResult<[SaxEventType, Detail]>;
  while (openTags >= 0 && !(result = await eventIter.next()).done) {
    const [evt, data] = result.value;
    if (evt === SaxEventType.CloseTag) {
      openTags--;
    } else if (evt !== SaxEventType.OpenTag) {
      continue;
    }

    openTags++;
    const tag = data as Tag;
    if (tag.name === 'TextBlock') {
      await handleBlock(eventIter, ctx, data as Tag);
    }
  }
  ctx.elementStack.pop();
  for (const elem of ['blocks', 'paragraphs', 'lines'] as const) {
    if ((page[elem] ?? []).some((e) => e.polygon !== undefined)) {
      page.features.push('POLYGONS');
    }
  }
  if (page.words.some((w) => w.choices !== undefined)) {
    page.features.push('CHOICES');
  }
  if (page.words.some((w) => w.hyphenStart)) {
    page.features.push('HYPHEN');
  }
  if (page.words.some((w) => w.confidence !== undefined)) {
    page.features.push('CONFIDENCE');
  }
  if (ctx.sourceImageInformation) {
    page.imageSource = ctx.sourceImageInformation;
    // Source Image information only applies to the first page in a file, all further pages
    // are undefined, as far as I can tell from the ALTO XSD.
    ctx.sourceImageInformation = undefined;
  }
  return page;
}

async function handleBlock(
  eventIter: AsyncIterableIterator<[SaxEventType, Detail]>,
  ctx: ParserContext,
  tag: Tag,
): Promise<void> {
  const dims = getCoordinates(tag);
  const block = makeBlock({
    x: dims.x ? dims.x! * ctx.scaleFactor : -1,
    y: dims.y ? dims.y! * ctx.scaleFactor : -1,
    width: dims.width ? dims.width * ctx.scaleFactor : -1,
    height: dims.height ? dims.height * ctx.scaleFactor : -1,
  });
  (ctx.elementStack.slice(-1)[0] as OcrPage).children.push(block);
  ctx.elementStack.push(block);

  let openTags = 0;
  let result: IteratorResult<[SaxEventType, Detail]>;
  while (openTags >= 0 && !(result = await eventIter.next()).done) {
    const [evt, data] = result.value;
    if (evt === SaxEventType.CloseTag) {
      openTags--;
    } else if (evt !== SaxEventType.OpenTag) {
      continue;
    }

    openTags++;
    const tag = data as Tag;
    if (tag.name === 'TextLine') {
      await handleLine(eventIter, ctx, data as Tag);
    } else if (tag.name === 'Shape') {
      const polygon = await handleShape(eventIter, ctx, tag);
      if (polygon != null) {
        block.polygon = polygon;
      }
    }
  }
  ctx.elementStack.pop();
}

async function handleLine(
  eventIter: AsyncIterableIterator<[SaxEventType, Detail]>,
  ctx: ParserContext,
  tag: Tag,
): Promise<void> {
  const dims = getCoordinates(tag);
  const line = makeLine({
    x: dims.x ? dims.x * ctx.scaleFactor : -1,
    y: dims.y ? dims.y * ctx.scaleFactor : -1,
    width: dims.width ? dims.width * ctx.scaleFactor : -1,
    height: dims.height ? dims.height * ctx.scaleFactor : -1,
  });
  (ctx.elementStack.slice(-1)[0] as OcrBlock).children.push(line);
  ctx.elementStack.push(line);

  let openTags = 0;
  let result: IteratorResult<[SaxEventType, Detail]>;
  while (openTags >= 0 && !(result = await eventIter.next()).done) {
    const [evt, data] = result.value;
    if (evt === SaxEventType.CloseTag) {
      openTags--;
    }
    if (evt !== SaxEventType.OpenTag) {
      continue;
    }

    openTags++;
    const tag = data as Tag;
    if (tag.name === 'String') {
      await handleWord(eventIter, ctx, data as Tag);
      // Words handle their closing themselves
      openTags--;
    } else if (tag.name === 'SP') {
      line.children.push(' ');
    } else if (tag.name === 'Shape') {
      const polygon = await handleShape(eventIter, ctx, tag);
      if (polygon != null) {
        line.polygon = polygon;
      }
    }
  }
  ctx.elementStack.pop();

  // Check if spaces are explicitely encoded
  if (!line.children.find((c) => c === ' ')) {
    // Add spaces between words, except for last word on line
    line.children = line.children.flatMap((c, i) =>
      i === line.children.length - 1 ? [c] : [c, ' '],
    );
  }
}

async function handleShape(
  eventIter: AsyncIterableIterator<[SaxEventType, Detail]>,
  ctx: ParserContext,
  tag: Tag,
): Promise<Polygon | null> {
  let polygon: Polygon | null = null;
  let openTags = 0;
  let result: IteratorResult<[SaxEventType, Detail]>;
  while (openTags >= 0 && !(result = await eventIter.next()).done) {
    const [evt, data] = result.value;
    if (evt === SaxEventType.CloseTag) {
      openTags--;
    }
    if (evt !== SaxEventType.OpenTag) {
      continue;
    }
    openTags++;
    const tag = data as Tag;
    if (tag.name === 'Polygon') {
      const points = getAttributeValue(tag, 'POINTS');
      if (!points) {
        logWithPosition(
          '<Polygon> without POINTS attribute, ignoring',
          'warn',
          tag,
        );
        continue;
      }
      polygon = points
        .split(' ')
        .map((p) => p.split(/(,| )/).map((c) => Number.parseFloat(c)))
        .reduce((acc, [c]) => {
          const lastElem: number[] | undefined = acc.slice(-1)[0];
          if (!lastElem || lastElem.length === 2) {
            acc.push([c]);
          } else {
            lastElem.push(c);
          }
          return acc;
        }, [] as number[][])
        .map(([x, y]) => ({ x: x * ctx.scaleFactor, y: y * ctx.scaleFactor }));
    }
  }
  const parent = ctx.elementStack.slice(-1)[0] as OcrBlock | OcrLine | OcrWord;
  if (parent.x < 0 || parent.y < 0 || parent.width < 0 || parent.height < 0) {
    if (polygon) {
      parent.x = Math.min(...polygon.map((p) => p.x));
      parent.y = Math.min(...polygon.map((p) => p.y));
      parent.width = Math.max(...polygon.map((p) => p.x)) - parent.x;
      parent.height = Math.max(...polygon.map((p) => p.y)) - parent.y;
    } else {
      throw new Error(
        `Could not get dimensions for ${parent.type}, no HPOS/VPOS/WIDTH/HEIGHT attribs and no associated Shape`,
      );
    }
  }
  return polygon;
}

async function handleWord(
  eventIter: AsyncIterableIterator<[SaxEventType, Detail]>,
  ctx: ParserContext,
  tag: Tag,
): Promise<void> {
  const dims = getCoordinates(tag);
  let text = getAttributeValue(tag, 'CONTENT');
  if (!text) {
    throw new Error('Could not get String text');
  }

  if (text.includes('&')) {
    text = resolveXmlEntities(text);
  }

  const word: OcrWord = {
    type: 'word',
    x: dims.x ? dims.x * ctx.scaleFactor : -1,
    y: dims.y ? dims.y * ctx.scaleFactor : -1,
    width: dims.width ? dims.width * ctx.scaleFactor : -1,
    height: dims.height ? dims.height * ctx.scaleFactor : -1,
    text,
  };
  ctx.elementStack.push(word);

  const subsType = getAttributeValue(tag, 'SUBS_TYPE');
  if (subsType === 'HypPart1') {
    word.hyphenStart = true;
  }

  const wordConfidence = getAttributeValue(tag, 'WC');
  if (wordConfidence !== undefined) {
    word.confidence = Number.parseFloat(wordConfidence);
  }

  let openTags = 0;
  let result: IteratorResult<[SaxEventType, Detail]>;
  while (openTags >= 0 && !(result = await eventIter.next()).done) {
    const [evt, data] = result.value;
    if (evt === SaxEventType.CloseTag) {
      openTags--;
      if ((data as Tag).name === 'ALTERNATIVE') {
        if (!('choices' in word)) {
          word.choices = [];
        }
        let altText = (data as Tag).textNodes
          .filter((t) => t.value.trim() !== '')
          .map((t) => t.value.trim())
          .join('');
        if (altText.includes('&')) {
          altText = resolveXmlEntities(altText);
        }
        word.choices!.push({
          text: altText,
        });
      }
    } else if (evt === SaxEventType.OpenTag) {
      openTags++;
      if ((data as Tag).name === 'Shape') {
        const polygon = await handleShape(eventIter, ctx, tag);
        openTags--;
        if (polygon != null) {
          word.polygon = polygon;
        }
      }
    }
  }
  const line = ctx.elementStack.slice(-2)[0] as OcrLine;
  line.children.push(word);
  ctx.elementStack.pop();
}

async function handleSourceImageInformation(
  eventIter: AsyncGenerator<[SaxEventType, Detail], any, unknown>,
  ctx: ParserContext,
  _tag: Tag,
): Promise<void> {
  let openTags = 0;
  let result: IteratorResult<[SaxEventType, Detail]>;
  ctx.sourceImageInformation = {};
  while (openTags >= 0 && !(result = await eventIter.next()).done) {
    const [evt, data] = result.value;
    const tag = data as Tag;
    if (evt === SaxEventType.CloseTag) {
      openTags--;
      if (tag.name === 'fileName') {
        ctx.sourceImageInformation.fileName = tag.textNodes
          .map((t) => t.value)
          .join('');
      } else if (tag.name === 'fileIdentifier') {
        let ident = tag.textNodes.map((t) => t.value).join('');
        const location = getAttributeValue(tag, 'fileIdentifierLocation');
        if (location) {
          ident = `${location}:${ident}`;
        }
        ctx.sourceImageInformation.fileIdentifier = ident;
      } else if (tag.name === 'documentIdentifier') {
        let ident = tag.textNodes.map((t) => t.value).join('');
        const location = getAttributeValue(tag, 'documentIdentifierLocation');
        if (location) {
          ident = `${location}:${ident}`;
        }
        ctx.sourceImageInformation.documentIdentifier = ident;
      }
    } else if (evt === SaxEventType.OpenTag) {
      openTags++;
    }
  }
}

/**
 * Parse an ALTO document.
 *
 * @param hocr The ALTO document to parse. Can be a `string`, a `Uint8Array` or
 *   a `ReadableStream`.
 * @param referenceSizes An array of dimensions for each page in the document,
 *   or a callback that returns the dimensions for a given page, based on its
 *   numerical index and its XML attributes. These dimensions can be used to
 *   scale all coordinates to a given reference frame in cases when the display
 *   resolution is different from the OCR resolution. A reference size might be
 *   required to obtain pixel coordinates, since in some documents the
 *   coordinates are expressed in non-pixel units.
 * @returns An async generator that yields `OcrPage` objects.
 */
export async function* parseAltoPages(
  alto: ReadableStream | Uint8Array | string,
  referenceSizes?: Dimensions[] | ReferenceSizeCallback,
): AsyncGenerator<OcrPage> {
  if (Array.isArray(referenceSizes)) {
    const sizeArr = referenceSizes;
    referenceSizes = (idx: number) => sizeArr[idx] ?? null;
  }
  const readable: ReadableStream<Uint8Array> =
    typeof alto === 'string' || alto instanceof Uint8Array
      ? toReadableStream(alto)
      : alto;
  const parser = await getParser(readable);
  const ctx: ParserContext = {
    scaleFactor: 1,
    elementStack: [],
  };
  const eventIter = parser.iterate();
  let result: IteratorResult<[SaxEventType, Detail]>;
  let pageIdx = 0;
  while (!(result = await eventIter.next()).done) {
    const [evt, data] = result.value;

    if (
      evt === SaxEventType.CloseTag &&
      (data as Tag).name === 'MeasurementUnit'
    ) {
      const unit = (data as Tag).textNodes[0].value.trim();
      if (unit !== 'pixel' && !referenceSizes) {
        throw new Error(`Measurement unit ${unit} requires a reference size.`);
      }
    }

    // Handlers advance the iterator themselves until they're done with their
    // respective element, this entry level just needs to recognize the
    // beginning of the page element and then hand over to the page handler
    // function.
    if (evt !== SaxEventType.OpenTag) {
      continue;
    }

    const tag = data as Tag;
    if (tag.name === 'Page') {
      const page = await handlePage(
        eventIter,
        ctx,
        data as Tag,
        referenceSizes?.(pageIdx, getAllAttributes(data as Tag)) ?? undefined,
      );
      if (page) {
        yield page;
        pageIdx++;
      }
    } else if (tag.name === 'sourceImageInformation') {
      await handleSourceImageInformation(eventIter, ctx, tag);
    }
  }
}
