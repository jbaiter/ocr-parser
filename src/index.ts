import { parseHocrPages } from './hocr.js';
import { parseAltoPages } from './alto.js';
import { toReadableStream, isInitialized } from './util.js';
import { type Dimensions, type OcrPage } from './model.js';

export { parseHocrPages } from './hocr.js';
export { parseAltoPages } from './alto.js';
export { initialize, isInitialized, setupLogging, SAX_WASM_VERSION } from './util.js';
export type {
  Dimensions,
  Coordinates,
  PageLocation,
  WordChoice,
  OcrPage,
  OcrBlock,
  OcrParagraph,
  OcrLine,
  OcrWord,
  OcrFeature,
  OcrFeatures,
  ImageSource,
  Polygon,
} from './model';

/**
 * Callback to provide a reference size for a given page.
 *
 * @param idx: 0-based index of the page in the document.
 * @param pageAttribs: XML attributes of the page element.
 * @returns The reference size , or `null` if no reference size is available.
 */
export type ReferenceSizeCallback = (
  idx: number,
  pageAttribs: { [key: string]: string },
) => Dimensions | null;

/**
 * Parse a single OCR page in either ALTO or hOCR format.
 *
 * Must have run {@link initialize} first.
 *
 * @param markup The document to parse. Can be a `string`, a `Uint8Array` or a
 *   `ReadableStream`.
 * @param format The format of the document.
 * @param referenceSizes Reference page dimensions These dimensions can be used
 *   to scale all coordinates to a given reference frame in cases when the
 *   display resolution is different from the OCR resolution. In the case of
 *   ALTO, a reference size might be required to obtain pixel coordinates,
 *   since in some documents the coordinates are expressed in non-pixel units.
 * @returns An async generator that yields `OcrPage` objects.
 */
export async function parseOcrPage(
  /** HOCR/ALTO markup for a single page. */
  markup: string | ReadableStream | Uint8Array,
  format: 'hocr' | 'alto',
  referenceSize?: Dimensions,
): Promise<OcrPage> {
  if (!isInitialized()) {
    throw new Error(
      'XML parser WASM not initialized, call initialize() first!',
    );
  }
  let page: OcrPage | null;
  if (format === 'hocr') {
    page = (
      await parseHocrPages(
        markup,
        referenceSize ? [referenceSize] : undefined,
      ).next()
    ).value;
  } else if (format === 'alto') {
    page = (
      await parseAltoPages(
        markup,
        referenceSize ? [referenceSize] : undefined,
      ).next()
    ).value;
  } else {
    throw new Error(`Unsupported format: ${format}`);
  }
  if (!page) {
    throw new Error('Failed to parse page');
  }
  return page;
}

/**
 * Parse all OCR pages in the given hOCR or ALTO markup.
 *
 * Must have run {@link initialize} first.
 *
 * @param markup The document to parse. Can be a `string`, a `Uint8Array` or a
 *   `ReadableStream`.
 * @param format The format of the document.
 * @param referenceSizes An array of dimensions for each page in the document,
 *   or a callback that returns the dimensions for a given page, based on its
 *   numerical index and its XML attributes.These dimensions can be used to
 *   scale all coordinates to a given reference frame in cases when the display
 *   resolution is different from the OCR resolution. In the case of ALTO, a
 *   reference size might be required to obtain pixel coordinates, since in
 *   some documents the coordinates are expressed in non-pixel units.
 * @returns An async generator that yields `OcrPage` objects.
 */
export async function* parseOcrPages(
  markup: string | ReadableStream | Uint8Array,
  format: 'hocr' | 'alto',
  referenceSizes?: Dimensions[] | ReferenceSizeCallback,
): AsyncGenerator<OcrPage> {
  if (!isInitialized()) {
    throw new Error(
      'XML parser WASM not initialized, call initialize() first!',
    );
  }
  if (!(markup instanceof ReadableStream)) {
    markup = toReadableStream(markup);
  }
  switch (format) {
    case 'hocr':
      yield* parseHocrPages(markup, referenceSizes);
      break;
    case 'alto':
      yield* parseAltoPages(markup, referenceSizes);
      break;
  }
}
