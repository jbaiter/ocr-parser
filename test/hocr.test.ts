import fs from 'fs';
import { describe, test, expect, beforeAll } from 'vitest';
import { parseHocrPages, initialize } from '../src';

// Custom serializer that ignores getters
expect.addSnapshotSerializer({
  test: (val) =>
    typeof val === 'object' &&
    Object.values(Object.getOwnPropertyDescriptors(val)).some((d) => d.get),
  serialize(val, config, indentation, depth, refs, printer) {
    const filtered = Object.fromEntries(
      Object.entries(val).filter(
        ([k]) => !Object.getOwnPropertyDescriptor(val, k)?.get,
      ),
    );
    return printer(filtered, config, indentation, depth, refs);
  },
});

//@ts-expect-error
import WITH_EMPTY_LINE from './__fixtures__/hocr_emptyline.html?raw';

const makeHocr = (hocrBody: string) => `
<?xml version="1.0" encoding="UTF-8"?>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
  <head>
    <title>Sample hOCR Document</title>
    <meta charset="UTF-8" />
    <meta name='ocr-system' content='tesseract 4.1.1' />
    <meta name='ocr-capabilities' content='ocr_page ocr_carea ocr_par ocr_line ocrx_word' />
  </head>
  <body>
  ${hocrBody}
  </body>
</html>
`;
const BASIC_HOCR = makeHocr(`
  <div class='ocr_page' id='page_1' title='image "sample.jpg"; bbox 0 0 800 600; ppageno 0'>
    <div class='ocr_carea' id='block_1_1' title="bbox 100 200 500 400">
      <p class='ocr_par' id='par_1_1' lang='eng'>
        <span class='ocr_line' id='line_1_1' title="bbox 100 200 500 220; baseline -0.053 -2">
          <span class='ocrx_word' id='word_1_1' title="bbox 100 200 150 220; x_wconf 94">Hello</span>
          <span class='ocrx_word' id='word_1_2' title="bbox 200 200 250 220; x_wconf 95">World,</span>
          <span class='ocrx_word' id='word_1_3' title="bbox 260 200 310 220; x_wconf 96">how</span>
          <span class='ocrx_word' id='word_1_4' title="bbox 320 200 370 220; x_wconf 97">are</span>
          <span class='ocrx_word' id='word_1_5' title="bbox 380 200 430 220; x_wconf 98">the</span>
          <span class='ocrx_word' id='word_1_5' title="bbox 400 200 450 220; x_wconf 98">Nightin\u00AD</span>
        </span>
        <span class="ocr_line" id="line_1_2" title="bbox 100 260 500 280; baseline -0.053 -2">
          <span class='ocrx_word' id='word_1_1' title="bbox 100 280 150 300; x_wconf 94">gales</span>
        </span>
      </p>
    </div>
  </div>
`);
const HOCR_WITH_ALTERNATIVES = makeHocr(`
    <div class="ocr_page" id="ID1" title="bbox 0 0 20145 26970; ppageno 0">
      <div class="ocrx_block" title="bbox 732 246 19674 1488">
        <span class="ocr_line" title="bbox 19326 246 19674 750">
          <span class="ocrx_word" title="bbox 19326 246 19674 750;x_conf 80.95">i</span>
        </span>
        <span class="ocr_line" title="bbox 744 720 795 1017">
          <span class="ocrx_word" title="bbox 744 720 795 1017;x_conf 95.24">IIi</span>
        </span>
        <span class="ocr_line" title="bbox 732 573 1344 1485">
          <span class="ocrx_word" title="bbox 732 846 795 1482;x_conf 95.24">IiI</span>
          <span class="ocrx_word" title="bbox 1272 573 1344 1485;x_conf 80.95">j</span>
        </span>
        <span class="ocr_line" title="bbox 732 1026 762 1482">
          <span class="ocrx_word" title="bbox 732 1026 762 1482;x_conf 80.95">I</span>
        </span>
        <span class="ocr_line" title="bbox 1200 1236 19398 1488">
          <span class="ocrx_word" title="bbox 1200 1236 1230 1488;x_conf 80.95">I</span>
          <span class="ocrx_word" title="bbox 1380 1242 1398 1374;x_conf 80.95">i</span>
          <span class="ocrx_word" title="bbox 2148 1314 2232 1452;x_conf 80.95">II</span>
          <span class="ocrx_word" title="bbox 7878 1305 8211 1422;x_conf 100.00">THE</span>
          <span class="ocrx_word" title="bbox 8373 1305 9267 1425;x_conf 100.00">STANDARD</span>
          <span class="ocrx_word" title="bbox 9504 1299 10068 1416;x_conf 100.00">OGDEN</span>
          <span class="ocrx_word" title="bbox 10254 1308 10722 1428;x_conf 95.24">
            <span class="alternatives">
              <ins class="alt">UTAH</ins>
              <del class="alt">UTAR</del>
              <del class="alt">VTA</del>
            </span>
          </span>
          <span class="ocrx_word" title="bbox 10917 1305 11820 1425;x_conf 100.00">SATURDAY</span>
          <span class="ocrx_word" title="bbox 12012 1302 12669 1440;x_conf 100.00">AUGUST</span>
          <span class="ocrx_word" title="bbox 12822 1326 12903 1446;x_conf 100.00">7</span>
          <span class="ocrx_word" title="bbox 13095 1320 13401 1446;x_conf 96.83">
            <span class="alternatives">
              <ins class="alt">1909</ins>
              <del class="alt">1900</del>
            </span>
          </span>
          <span class="ocrx_word" title="bbox 19326 1272 19398 1488;x_conf 80.95">j</span></span>
      </div>
    </div>
`);

const saxWasm = fs.readFileSync(
  __dirname + '/../node_modules/sax-wasm/lib/sax-wasm.wasm',
);

async function toArray<T>(gen: AsyncGenerator<T>): Promise<T[]> {
  const result: T[] = [];
  for await (let page of gen) {
    result.push(page);
  }
  return result;
}

describe('hOCR parser', () => {
  beforeAll(() => {
    initialize(() => Promise.resolve(new Uint8Array(saxWasm)));
  });
  test('should parse a simple hOCR document', async () => {
    const pages = await toArray(
      parseHocrPages(BASIC_HOCR, [{ width: 800, height: 600 }]),
    );
    expect(pages).toHaveLength(1);
    const page = pages[0];
    expect(page).toMatchSnapshot();
    expect(page?.text).toEqual('Hello World, how are the Nightingales');
  });

  test('should be able to handle a hOCR doc with alternatives', async () => {
    const pages = await toArray(
      parseHocrPages(HOCR_WITH_ALTERNATIVES, [
        {
          width: 20145,
          height: 26970,
        },
      ]),
    );
    expect(pages).toHaveLength(1);
    const page = pages[0];
    expect(page?.text).toMatchInlineSnapshot(
      '"i IIi IiI j I I i II THE STANDARD OGDEN UTAH SATURDAY AUGUST 7 1909 j"',
    );
    expect(page?.lines).toMatchSnapshot();
  });

  test('should be able to handle lines with only empty words', async () => {
    const pages = await toArray(parseHocrPages(WITH_EMPTY_LINE));
    expect(pages).toHaveLength(1);
    expect(pages[0].lines).toMatchSnapshot();
  });
});
