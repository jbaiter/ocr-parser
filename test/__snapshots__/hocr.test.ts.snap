// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hOCR parser > should be able to handle a hOCR doc with alternatives 2`] = `
[
  {
    "children": [
      {
        "height": 504,
        "hyphenStart": false,
        "text": "i",
        "type": "word",
        "width": 348,
        "x": 19326,
        "y": 246,
      },
    ],
    "height": 504,
    "type": "line",
    "width": 348,
    "x": 19326,
    "y": 246,
  },
  {
    "children": [
      {
        "height": 297,
        "hyphenStart": false,
        "text": "IIi",
        "type": "word",
        "width": 51,
        "x": 744,
        "y": 720,
      },
    ],
    "height": 297,
    "type": "line",
    "width": 51,
    "x": 744,
    "y": 720,
  },
  {
    "children": [
      {
        "height": 636,
        "hyphenStart": false,
        "text": "IiI",
        "type": "word",
        "width": 63,
        "x": 732,
        "y": 846,
      },
      " ",
      {
        "height": 912,
        "hyphenStart": false,
        "text": "j",
        "type": "word",
        "width": 72,
        "x": 1272,
        "y": 573,
      },
    ],
    "height": 912,
    "type": "line",
    "width": 612,
    "x": 732,
    "y": 573,
  },
  {
    "children": [
      {
        "height": 456,
        "hyphenStart": false,
        "text": "I",
        "type": "word",
        "width": 30,
        "x": 732,
        "y": 1026,
      },
    ],
    "height": 456,
    "type": "line",
    "width": 30,
    "x": 732,
    "y": 1026,
  },
  {
    "children": [
      {
        "height": 252,
        "hyphenStart": false,
        "text": "I",
        "type": "word",
        "width": 30,
        "x": 1200,
        "y": 1236,
      },
      " ",
      {
        "height": 132,
        "hyphenStart": false,
        "text": "i",
        "type": "word",
        "width": 18,
        "x": 1380,
        "y": 1242,
      },
      " ",
      {
        "height": 138,
        "hyphenStart": false,
        "text": "II",
        "type": "word",
        "width": 84,
        "x": 2148,
        "y": 1314,
      },
      " ",
      {
        "height": 117,
        "hyphenStart": false,
        "text": "THE",
        "type": "word",
        "width": 333,
        "x": 7878,
        "y": 1305,
      },
      " ",
      {
        "height": 120,
        "hyphenStart": false,
        "text": "STANDARD",
        "type": "word",
        "width": 894,
        "x": 8373,
        "y": 1305,
      },
      " ",
      {
        "height": 117,
        "hyphenStart": false,
        "text": "OGDEN",
        "type": "word",
        "width": 564,
        "x": 9504,
        "y": 1299,
      },
      " ",
      {
        "choices": [
          {
            "text": "UTAR",
          },
          {
            "text": "VTA",
          },
        ],
        "height": 120,
        "hyphenStart": false,
        "text": "UTAH",
        "type": "word",
        "width": 468,
        "x": 10254,
        "y": 1308,
      },
      " ",
      {
        "height": 120,
        "hyphenStart": false,
        "text": "SATURDAY",
        "type": "word",
        "width": 903,
        "x": 10917,
        "y": 1305,
      },
      " ",
      {
        "height": 138,
        "hyphenStart": false,
        "text": "AUGUST",
        "type": "word",
        "width": 657,
        "x": 12012,
        "y": 1302,
      },
      " ",
      {
        "height": 120,
        "hyphenStart": false,
        "text": "7",
        "type": "word",
        "width": 81,
        "x": 12822,
        "y": 1326,
      },
      " ",
      {
        "choices": [
          {
            "text": "1900",
          },
        ],
        "height": 126,
        "hyphenStart": false,
        "text": "1909",
        "type": "word",
        "width": 306,
        "x": 13095,
        "y": 1320,
      },
      " ",
      {
        "height": 216,
        "hyphenStart": false,
        "text": "j",
        "type": "word",
        "width": 72,
        "x": 19326,
        "y": 1272,
      },
    ],
    "height": 252,
    "type": "line",
    "width": 18198,
    "x": 1200,
    "y": 1236,
  },
]
`;

exports[`hOCR parser > should be able to handle lines with only empty words 1`] = `
[
  {
    "children": [],
    "height": 1225,
    "type": "line",
    "width": 771,
    "x": 0,
    "y": 0,
  },
  {
    "children": [],
    "height": 1,
    "type": "line",
    "width": 1,
    "x": 0,
    "y": 0,
  },
]
`;

exports[`hOCR parser > should parse a simple hOCR document 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "baseline": [
                {
                  "x": 100,
                  "y": 198,
                },
                {
                  "x": 500,
                  "y": 176.8,
                },
              ],
              "children": [
                {
                  "confidence": 0.94,
                  "height": 20,
                  "hyphenStart": false,
                  "text": "Hello",
                  "type": "word",
                  "width": 50,
                  "x": 100,
                  "y": 200,
                },
                " ",
                {
                  "confidence": 0.95,
                  "height": 20,
                  "hyphenStart": false,
                  "text": "World,",
                  "type": "word",
                  "width": 50,
                  "x": 200,
                  "y": 200,
                },
                " ",
                {
                  "confidence": 0.96,
                  "height": 20,
                  "hyphenStart": false,
                  "text": "how",
                  "type": "word",
                  "width": 50,
                  "x": 260,
                  "y": 200,
                },
                " ",
                {
                  "confidence": 0.97,
                  "height": 20,
                  "hyphenStart": false,
                  "text": "are",
                  "type": "word",
                  "width": 50,
                  "x": 320,
                  "y": 200,
                },
                " ",
                {
                  "confidence": 0.98,
                  "height": 20,
                  "hyphenStart": false,
                  "text": "the",
                  "type": "word",
                  "width": 50,
                  "x": 380,
                  "y": 200,
                },
                " ",
                {
                  "confidence": 0.98,
                  "height": 20,
                  "hyphenStart": true,
                  "text": "Nightin",
                  "type": "word",
                  "width": 50,
                  "x": 400,
                  "y": 200,
                },
              ],
              "height": 20,
              "type": "line",
              "width": 400,
              "x": 100,
              "y": 200,
            },
            {
              "baseline": [
                {
                  "x": 100,
                  "y": 258,
                },
                {
                  "x": 500,
                  "y": 236.8,
                },
              ],
              "children": [
                {
                  "confidence": 0.94,
                  "height": 20,
                  "hyphenStart": false,
                  "text": "gales",
                  "type": "word",
                  "width": 50,
                  "x": 100,
                  "y": 280,
                },
              ],
              "height": 20,
              "type": "line",
              "width": 400,
              "x": 100,
              "y": 260,
            },
          ],
          "height": undefined,
          "type": "paragraph",
          "width": undefined,
          "x": undefined,
          "y": undefined,
        },
      ],
      "height": 200,
      "type": "block",
      "width": 400,
      "x": 100,
      "y": 200,
    },
  ],
  "features": [
    "HYPHEN",
    "CONFIDENCE",
  ],
  "height": 600,
  "imageSource": {
    "fileName": "sample.jpg",
  },
  "physicalPageNumber": 0,
  "type": "page",
  "width": 800,
}
`;
